File: Python/Question Practice Python/Singly_Linked_Lists.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insertAtEnd(self, val):
        newNode = Node(val)
        if self.head is None:
            self.head = newNode
            print(f"Inserted head node with value {newNode.data}")
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next

        temp.next = newNode
        print(f"Inserted new node with value {newNode.data}")

    def insertAtBeginning(self, val):
        newNode = Node(val)
        newNode.next = self.head
        self.head = newNode
        print(f"Inserted head node with value {newNode.data}")

    def deleteAtBeginning(self):
        if self.head is None:
            print("List is empty")
            return
        
        temp = self.head
        self.head = self.head.next
        print(f"Deleted node with value {temp.data}")
        del temp

    def deleteAtEnd(self):
        if self.head is None:
            print("List is empty")
            return
        
        if self.head.next is None:
            print(f"Deleted node with value {self.head.data}")
            del self.head
            self.head = None
            return
        
        temp = self.head
        while temp.next.next is not None:
            temp = temp.next

        toDelete = temp.next
        temp.next = None
        print(f"Deleted node with value {toDelete.data}")
        del toDelete

    def display(self):
        temp = self.head
        while temp is not None:
            print(f"{temp.data} -> ", end="")
            temp = temp.next

        print("NULL")

list = LinkedList()
print("Linked List Operations:")
print("1. Insert at end\n2. Insert at beginning\n3. Delete at beginning\n4. Delete at end\n5. Display")

while True:
    oper = int(input("\nEnter the Operation: "))
    if oper == 1:
        val = int(input("Enter the value to insert: "))
        list.insertAtEnd(val)
        
    elif oper == 2:
        val = int(input("Enter the value to insert: "))
        list.insertAtBeginning(val)
        
    elif oper == 3:
        list.deleteAtBeginning()
        
    elif oper == 4:
        list.deleteAtEnd()
        
    elif oper == 5:
        list.display()
        
    else:
        print("Invalid Operation")

File: C/Question Practice C/Singly_Linked_Lists.c
#include<stdio.h>
#include<stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct LinkedList {struct Node* head;};
void Initialize(struct LinkedList* list) {list->head = NULL;}

void InsertAtBeginning(struct LinkedList* list, int val) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = val;
    newNode->next = list->head;
    list->head = newNode;
    printf("Inserted head node with value %d\n", newNode->data);
}

void InsertAtEnd(struct LinkedList* list, int val) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = val;
    newNode->next = NULL;

    struct Node* temp = list->head;
    if (temp == NULL) {
        list->head = newNode;
        printf("Inserted head node with value %d\n", newNode->data);
        return;
    }

    while (temp->next != NULL) temp = temp->next;
    temp->next = newNode;
    printf("Inserted new node with value %d\n", newNode->data);
}

void DeleteAtBeginning(struct LinkedList* list){
    if (list->head == NULL){
        printf("List is empty\n");
        return;
    }

    struct Node* temp = list->head;
    list->head = list->head->next;
    printf("Deleted node with value %d\n", temp->data);
    free(temp);
}

void Searching(struct LinkedList* list, int val){
    int found = 0, node = 0;
    struct Node* temp = list->head;

    while (temp != NULL){
        node++;
        if (temp->data == val){
            found = 1;
            break;
        }
        temp = temp->next;
    }

    if (!found) printf("Value %d not found in the list\n", val);
    else printf("Value %d found at node %d\n", val, node);
}

void DeleteAtEnd(struct LinkedList* list){
    if (list->head == NULL){
        printf("List is empty\n");
        return;
    }

    struct Node* temp = list->head;
    if (temp->next == NULL) {
        printf("Deleted head node with value %d\n", temp->data);
        free(temp);
        list->head = NULL;
        return;
    }

    while (temp->next->next != NULL) temp = temp->next;
    struct Node* toDelete = temp->next;
    printf("Deleted node with value %d\n", toDelete->data);
    temp->next = NULL;
    free(toDelete);
}

void Display(struct LinkedList* list){
    struct Node* temp = list->head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

int main(){
    struct LinkedList l1;
    Initialize(&l1);

    printf("Linked List Operations:\n1. InsertAtBeginning\n2. InsertAtEnd\n3. DeleteAtBeginning\n4. DeleteAtEnd\n5. Searching\n6. Display\n");
    while (1){
        int choice, elem;
        printf("\nEnter Linked List Operation: ");
        scanf("%d",&choice);

        if (choice == 1) {
            printf("Enter the Value to insert: ");
            scanf("%d",&elem);
            InsertAtBeginning(&l1, elem); 
        } else if (choice == 2) {
            printf("Enter the Value to insert: ");
            scanf("%d",&elem);
            InsertAtEnd(&l1, elem);
        } 
        else if (choice == 3) DeleteAtBeginning(&l1);
        else if (choice == 4) DeleteAtEnd(&l1);
        else if (choice == 5) {
            printf("Enter Value to be Searched: ");
            scanf("%d", &elem);
            Searching(&l1, elem);
        } else if (choice == 6) Display(&l1);
        else printf("Enter Valid Operation\n");
    }
    
    return 0;
}